parl/remote/grpc_heartbeat/tests/heartbeat_server_arguments_test.py

```diff
@@ -0,0 +1,207 @@
+        assert not heartbeat_server_thread.is_alive()
+        assert not heartbeat_client_thread.is_alive()
+
+        assert self.server_exited == False  # the heartbeat server cannot exit normally
+        assert self.client_exited == True
+
+    def test_heartbeat_server_exit_with_kwargs(self):
+        arg1_value = 10
+
+        def server_exit_func(arg1):
+            print("exit heartbeat server")
+            assert arg1 == arg1_value
+            self.server_exited = True
+
+        heartbeat_server_thread = HeartbeatServerThread(
+            server_exit_func, exit_func_kwargs={"arg1": arg1_value})
+        heartbeat_server_thread.start()
+
+        server_address = heartbeat_server_thread.get_address()
+
+        def client_exit_func():
+            print("exit heartbeat client")
+            self.client_exited = True
+
+        heartbeat_client_thread = HeartbeatClientThread(
+            server_address, client_exit_func)
+        heartbeat_client_thread.start()
+
+        time.sleep(remote_constants.HEARTBEAT_RCVTIMEO_S * 2)
+
+        # check server and client are still alive after HEARTBEAT_RCVTIMEO_S * 2
+        assert heartbeat_server_thread.is_alive()
+        assert heartbeat_client_thread.is_alive()
+
+        heartbeat_server_thread.exit()  # manually exit the server
+
+        # wait for threads exiting
+        for _ in range(6):
+            if not heartbeat_server_thread.is_alive(
+            ) and not heartbeat_client_thread.is_alive():
+                break
+            time.sleep(remote_constants.HEARTBEAT_INTERVAL_S)
+
+        # check heartbeat server and client are exited
+        assert not heartbeat_server_thread.is_alive()
+        assert not heartbeat_client_thread.is_alive()
+
+        assert self.server_exited == True
+        assert self.client_exited == True
+
+    def test_heartbeat_server_exit_with_wrong_kwargs(self):
+        arg1_value = 10
+
+        def server_exit_func(arg1):
+            print("exit heartbeat server")
+            assert arg1 == arg1_value
+            self.server_exited = True
+
+        heartbeat_server_thread = HeartbeatServerThread(
+            server_exit_func, exit_func_kwargs={"wrong_args": arg1_value})
+        heartbeat_server_thread.start()
+
+        server_address = heartbeat_server_thread.get_address()
+
+        def client_exit_func():
+            print("exit heartbeat client")
+            self.client_exited = True
+
+        heartbeat_client_thread = HeartbeatClientThread(
+            server_address, client_exit_func)
+        heartbeat_client_thread.start()
+
+        time.sleep(remote_constants.HEARTBEAT_RCVTIMEO_S * 2)
+
+        # check server and client are still alive after HEARTBEAT_RCVTIMEO_S * 2
+        assert heartbeat_server_thread.is_alive()
+        assert heartbeat_client_thread.is_alive()
+
+        heartbeat_server_thread.exit()  # manually exit the server
+        # will raise an exception in the backend thread
```

Comment:
This is a good test, but the main thread should know that something has gone wrong when calling the exit function.